### PR TITLE
[Azure MySQL] create feature - phrase I

### PR DESCRIPTION
**Pull request by shenqianjin**
Basic Mode:
image

Advanced Mode:
image

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                    .type(AzureValidationInfo.Type.ERROR).build();
+        }
+        // validate special character
+        if (!StringUtils.isAlphanumeric(value)) {
+            final AzureValidationInfo.AzureValidationInfoBuilder builder = AzureValidationInfo.builder();
+            return builder.input(this).message("Admin username must only contain characters and numbers.").type(AzureValidationInfo.Type.ERROR).build();
+        }
+        // validate special admin username
+        if (StringUtils.equalsIgnoreCase(value, "azure_superuser")
+                || StringUtils.equalsIgnoreCase(value, "admin")
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java:56]: use `StringUtils.equalsAnyIgnoreCase`?

Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java:56]: We may store the invalid values in a constant array

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/mysql/AzureMySQLConfig.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    private char[] confirmPassword;
+
+    private boolean allowAccessFromAzureServices;
+    private boolean allowAccessFromLocalMachine;
+
+    public static AzureMySQLConfig getDefaultAzureMySQLConfig() {
+        final String defaultNameSuffix = DateFormatUtils.format(new Date(), "yyyyMMddHHmmss");
+        final AzureMySQLConfig config = new AzureMySQLConfig();
+        List<Subscription> selectedSubscriptions = AzureMvpModel.getInstance().getSelectedSubscriptions();
+        if (CollectionUtils.isNotEmpty(selectedSubscriptions)) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/mysql/AzureMySQLConfig.java:65]: Will `selectedSubscriptions` be empty? And do we need to handle this cornet case?

Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/mysql/AzureMySQLConfig.java:65]: it's better to check to confirm a list is not empty before we call xxxList.get(0).

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/AzurePasswordFieldInput.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        this.delegate = delegate;
+        this.passwordInitialized = passwordInitialized;
+    }
+
+    @Override
+    public AzureValidationInfo doValidate() {
+        if (!isPasswordInitialized()) {
+            return AzureValidationInfo.UNINITIALIZED;
+        }
+        return AzureFormInputComponent.super.doValidate();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/AzurePasswordFieldInput.java:54]: super.doValidate();?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import com.microsoft.azure.toolkit.intellij.common.ValidationDebouncedTextInput;
+import com.microsoft.azure.toolkit.lib.common.form.AzureValidationInfo;
+import com.microsoft.azuretools.azurecommons.helpers.NotNull;
+import lombok.Getter;
+import lombok.Setter;
+import org.apache.commons.lang3.StringUtils;
+
+public class AdminUsernameTextField extends ValidationDebouncedTextInput {
+
+    private static final int LENGTH_MIN = 1;
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java:33]: MIN_LENGTH

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    private static final int LENGTH_MAX = 16;
+    private static final String LENGTH_INVALILD_MESSAGE = "Admin username must be at least 1 characters and at most 16 characters.";
+    private static final String ALPHANUMBERIC_INVALID_MESSAGE = "Admin username must only contain characters and numbers.";
+    private static final String[] FIXED_NAMES_INVALID = new String[]{"azure_superuser", "admin", "administrator", "root", "guest", "public"};
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java:37]: INVALID_USER_NAMES

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/AzurePasswordFieldInput.java (hunk 3 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/AzurePasswordFieldInput.java:54]: AzureFormInputComponent is an interface. "supper" is not supported for it.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    private static final String FIXED_NAMES_INVALID_MESSAGE =
+            "Admin login name cannot be 'azure_superuser', 'admin', 'administrator', 'root', 'guest' or 'public'.";
+
+    @Getter
+    @Setter
+    private boolean passwordInitialized;
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java:43]: where to change passwordInitialized?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    public AzureValidationInfo doValidateValue() {
+        final AzureValidationInfo info = super.doValidateValue();
+        if (!AzureValidationInfo.OK.equals(info)) {
+            return info;
+        }
+        final String value = this.getValue();
+        // validate length
+        if (StringUtils.length(value) < 3 || StringUtils.length(value) > 63) {
+            final AzureValidationInfo.AzureValidationInfoBuilder builder = AzureValidationInfo.builder();
+            return builder.input(this).message("Server name must be at least 3 characters and at most 63 characters.")
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java:70]: return AzureValidationInfo.builder().input...

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java (hunk 6 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/AdminUsernameTextField.java:43]: The create dialog of mysql. will included in the next PR.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        // validate special character
+        if (!pattern.matcher(value).matches()) {
+            final AzureValidationInfo.AzureValidationInfoBuilder builder = AzureValidationInfo.builder();
+            return builder.input(this)
+                    .message("Server name must only contain lowercase letters, numbers, and hyphens. The server name must not start or end in a hyphen.")
+                    .type(AzureValidationInfo.Type.ERROR).build();
+        }
+        // validate availability
+        try {
+            Server server = MySQLMvpModel.findServer(subscription.subscriptionId(), resourceGroup.name(), this.getValue());
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java:82]: the name confliction check is wrong, you need to special check name availability api.

Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java:82]: you should check the name outside this subs for global exclusion.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            if (server != null) {
+                final AzureValidationInfo.AzureValidationInfoBuilder builder = AzureValidationInfo.builder();
+                return builder.input(this).message(this.getValue() + " already existed.").type(AzureValidationInfo.Type.ERROR).build();
+            }
+        } catch (NullPointerException e) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java:87]: which line may throw NPE? it is not good to catch NPE

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java (hunk 8 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java:82]: https://github.com/Azure/azure-libraries-for-java/blob/master/azure-mgmt-appservice/src/main/java/com/microsoft/azure/management/appservice/CheckNameResourceTypes.java

Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java:82]: image

Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java:82]: No API for MySQL. if a server name does not exist, SDK find method will return NPE.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java (hunk 9 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/ServerNameTextField.java:87]: No API for MySQL. if a server name does not exist, SDK find method will return NPE.

**Review by Flanker32 — APPROVED** [review]

**Pull request merged**